### ISSUE TITLE
Error when resuming training from checkpoint – training stops instead of continuing

### ISSUE THREAD

**Issue by kartikeykushwaha14** [k2-fsa/OmniVoice#117]
### Checks

- [x] This template is only for bug reports, usage problems go with 'Help Wanted'.
- [x] I have thoroughly reviewed the project documentation but couldn't find information to solve my problem.
- [x] I have searched for existing issues, including closed ones, and couldn't find a solution.
- [x] I am using English to submit this issue to facilitate community communication.

### Environment Details

Environment
OS: Ubuntu 22.04
Python version: 3.10
GPU: H200

I am facing an issue while trying to resume training from a saved checkpoint.

I trained my model up to 10,000 steps and a checkpoint was successfully saved. However, when I try to resume training from that checkpoint, the training does not continue. Instead, it immediately exits after loading and saving again the same checkpoint. 

please tell me how to solve this error 

### Steps to Reproduce

Steps to Reproduce
Start training normally
Let training run until a checkpoint is saved (e.g., step 10,000)
Restart training using the checkpoint (resume option)
Observe that training exits instead of continuing

{
    "llm_name_or_path": "Qwen/Qwen3-0.6B",
    "audio_vocab_size": 1025,
    "audio_mask_id": 1024,
    "num_audio_codebook": 8,

    "audio_codebook_weights": [8, 8, 6, 6, 4, 4, 2, 2],
    "drop_cond_ratio": 0.1,
    "prompt_ratio_range": [0.0, 0.3],
    "mask_ratio_range": [0.0, 1.0],
    "language_ratio": 0.8,
    "use_pinyin_ratio": 0.0,
    "instruct_ratio": 0.0,
    "only_instruct_ratio": 0.0,

    "resume_from_checkpoint": "TTS/OmniVoice_V2/3_lang_low_learning_rate_backup/checkpoint-10000",
    "init_from_checkpoint": null,

    "learning_rate": 2e-5,
    "weight_decay": 0.01,
    "max_grad_norm": 1.0,
    "steps": 20000,
    "seed": 42,
    "warmup_type": "ratio",
    "warmup_ratio": 0.01,
    "warmup_steps": 0,

    "batch_tokens": 8192,
    "gradient_accumulation_steps": 1,
    "num_workers": 2,

    "mixed_precision": "no",
    "allow_tf32": true,

    "logging_steps": 50,
    "eval_steps": 1000,
    "save_steps": 2000,
    "keep_last_n_checkpoints": -1
}

and using same file which is provided in cli folder train.py 

### ✔️ Expected Behavior

Training should resume from the saved checkpoint (e.g., step 10,000) and continue further training normally.

### ❌ Actual Behavior

Checkpoint loads successfully
Training does not proceed
Process exits immediately without continuing training

**Comment by zhu-han** [comment]
Can you provide the detailed training log when resuming training?

**Comment by zhu-han** [comment]
I cannot reproduce this issue. One possible reason is that you did not use the correct training configuration when resuming training. If you confirm that you are using the right config, please provide the full training log and reopen this issue.

**Comment by kartikeykushwaha14** [comment]
Thanks for your response.

I am facing an issue while resuming training from a checkpoint. Although training resumes correctly in terms of steps, the model performance degrades compared to training using init_from_checkpoint.

Training Workflow
Step 1: Initial Training (0 → 10000 steps)
 
{
    "llm_name_or_path": "Qwen/Qwen3-0.6B",
    "audio_vocab_size": 1025,
    "audio_mask_id": 1024,
    "num_audio_codebook": 8,

    "audio_codebook_weights": [8, 8, 6, 6, 4, 4, 2, 2],
    "drop_cond_ratio": 0.0,
    "prompt_ratio_range": [0.0, 0.3],
    "mask_ratio_range": [0.0, 1.0],
    "language_ratio": 1.0,
    "use_pinyin_ratio": 0.0,
    "instruct_ratio": 0.0,
    "only_instruct_ratio": 0.0,

    "resume_from_checkpoint": null,
    "init_from_checkpoint": "k2-fsa/OmniVoice",

    "learning_rate": 2e-5,
    "weight_decay": 0.01,
    "max_grad_norm": 1.0,
    "steps": 10000,
    "seed": 42,
    "warmup_type": "ratio",
    "warmup_ratio": 0.01,
    "warmup_steps": 0,

    "batch_tokens": 8192,
    "gradient_accumulation_steps": 1,
    "num_workers": 4,

    "mixed_precision": "no",
    "allow_tf32": true,

    "logging_steps": 50,
    "eval_steps": 1000,
    "save_steps": 1000,
    "keep_last_n_checkpoints": -1
}

Train on my dataset for sanskrit language working fine model learning data perfectly and good result when start from training init checkpoint .

Step 2: Resume Training (10000 → 20000 steps)

{
"llm_name_or_path": "Qwen/Qwen3-0.6B",
"audio_vocab_size": 1025,
"audio_mask_id": 1024,
"num_audio_codebook": 8,

"audio_codebook_weights": [8, 8, 6, 6, 4, 4, 2, 2],
"drop_cond_ratio": 0.1,
"prompt_ratio_range": [0.0, 0.3],
"mask_ratio_range": [0.0, 1.0],
"language_ratio": 0.8,
"use_pinyin_ratio": 0.0,
"instruct_ratio": 0.0,
"only_instruct_ratio": 0.0,

"resume_from_checkpoint": "TTS/OmniVoice_V2/3_lang_low_learning_rate_backup/checkpoint-10000",
"init_from_checkpoint": null,

"learning_rate": 2e-5,
"weight_decay": 0.01,
"max_grad_norm": 1.0,
"steps": 20000,
"seed": 42,
"warmup_type": "ratio",
"warmup_ratio": 0.01,
"warmup_steps": 0,

"batch_tokens": 8192,
"gradient_accumulation_steps": 1,
"num_workers": 2,

"mixed_precision": "no",
"allow_tf32": true,

"logging_steps": 50,
"eval_steps": 1000,
"save_steps": 2000,
"keep_last_n_checkpoints": -1
}

After trying again carefully, the training does resume, but I am facing another issue:

Training successfully resumes from step 10000
However, model performance degrades after resuming
The results are noticeably worse compared to:
Training continuously (0 → 20000), OR
Restarting using init_from_checkpoint

Because of this, I often have to restart training from step 0, which is inefficient and wastes compute.

When I resume training from the checkpoint, the results are not as good compared to training using `init_from_checkpoint` (starting again from the dataset). Because of this, I often have to restart training from step 0, which is time-consuming and leads to loss of previous training progress.

Could you please clarify the correct way to resume training so that performance is consistent and continues improving from the checkpoint?

Is there anything specific I should take care of (e.g., optimizer state, scheduler, data loading, or config differences) when using `resume_from_checkpoint`?

Thanks again for your help.

**Comment by zhu-han** [comment]
When resuming training, the dataloader does not resume from its last state (because we are using an IterableDataset), so the performance will differ from training from scratch. I personally never used the resume training mode in my experiments. If you find that fine-tuning with init_from_checkpoint yields better results, you can use this method instead. This approach does not resume the optimizer state and has a warmup stage.

**Comment by kartikeykushwaha14** [comment]
Thank you for your response and for building such a great project. We can clearly see the effort and thought you’ve put into it—it’s really appreciated. Your work is impressive and very helpful for the community.
